File: app/src/main/java/knf/kuma/updater/UpdateChecker.java
package knf.kuma.updater;

import android.content.Context;
import android.os.AsyncTask;
import android.util.Log;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

import knf.kuma.BuildConfig;
import knf.kuma.commons.Network;

public class UpdateChecker {
    public static void check(final Context context, final CheckListener listener) {
        if (Network.isConnected() && !BuildConfig.BUILD_TYPE.equals("playstore"))
            AsyncTask.execute(() -> {
                try {
                    Document document = Jsoup.connect("https://raw.githubusercontent.com/jordyamc/UKIKU/master/version.num").get();
                    int n_code = Integer.parseInt(document.select("body").first().ownText().trim());
                    int o_code = context.getPackageManager().getPackageInfo(context.getPackageName(), 0).versionCode;
                    if (n_code > o_code) {
                        listener.onNeedUpdate(String.valueOf(o_code), String.valueOf(n_code));
                    } else {
                        Log.e("Version", "Up to date: " + o_code);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
    }

    public interface CheckListener {
        void onNeedUpdate(String o_code, String n_code);
    }
}

File: app/src/main/java/knf/kuma/updater/UpdateChecker.kt
package knf.kuma.updater

import android.util.Log
import androidx.core.content.pm.PackageInfoCompat
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.lifecycleScope
import knf.kuma.commons.Network
import knf.kuma.commons.isFullMode
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.jsoup.Jsoup

object UpdateChecker {
    fun check(context: FragmentActivity, listener: CheckListener) {
        if (Network.isConnected && isFullMode)
            context.lifecycleScope.launch(Dispatchers.IO) {
                try {
                    val document =
                        Jsoup.connect("https://raw.githubusercontent.com/jordyamc/UKIKU/master/version.num")
                            .get()
                    val nCode = Integer.parseInt(document.select("body").first().ownText().trim())
                    val oCode = PackageInfoCompat.getLongVersionCode(
                        context.packageManager.getPackageInfo(
                            context.packageName,
                            0
                        )
                    ).toInt()
                    if (nCode > oCode) {
                        delay(2000)
                        listener.onNeedUpdate(oCode.toString(), nCode.toString())
                    } else {
                        context.filesDir.listFiles()
                            ?.filter {
                                !it.isDirectory && it.name.startsWith("update") && it.name.endsWith(
                                    ".apk"
                                )
                            }
                                ?.forEach {
                                    it.delete()
                                }
                        Log.e("Version", "Up to date: $oCode")
                        listener.onUpdateNotRequired()
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                    listener.onUpdateNotRequired()
                }
            }
        else
            listener.onUpdateNotRequired()
    }

    interface CheckListener {
        fun onNeedUpdate(o_code: String, n_code: String)
        fun onUpdateNotRequired()
    }
}
